### PR TITLE
Unit traits exclusive

### PR DESCRIPTION
**Pull request by Kingofd**
Please, for your convenience, test it with a build, trait config file included. Do not include trait config in merge, tho, only for testing

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
You will need to update the schema to add the new keys.

File: data/core/macros/traits.cfg
Comment (knyghtmare) [inline data/core/macros/traits.cfg:146]: Why is trait strong requiring this?

File: data/core/macros/traits.cfg
Comment (knyghtmare) [inline data/core/macros/traits.cfg:184]: I am not exactly sure why the core traits file is modified to ensure the combination `intelligent,quick` is never picked.

File: data/core/macros/traits.cfg
Comment (knyghtmare) [inline data/core/macros/traits.cfg:222]: Is there anything negative about the combo `resilient, intelligent`?

**Review by knyghtmare — CHANGES_REQUESTED** [review]
I think there's some things which need to be clarified. Commented them separately.

**Comment by knyghtmare** [comment]
> Do not include trait config in merge, tho, only for testing

Oh I see. This is just testing to see if the following happen/not-happen.

I can build over this and test it in the weekend. So, can you name the expected observations to compare with?

**Comment by Kingofd** [comment]
It might be helpful to also look at the case of `strong` and `quick`
Since `strong` needs `quick`, it can only ever occur, that strong has been picked as the second trait, so always `quick, strong` never the other way round

File: src/units/unit.cpp
Comment (soliton-) [inline src/units/unit.cpp:848]: Doesn't look like we have an explicit rule about it but I would either write this on one line or use curly braces. The way it is now you might just add an indented statement and expect it to be behind the if...

**Comment by Pentarctagon** [comment]
Is this something that could have one or more WML unit tests added for it?

File: src/units/unit.cpp
Comment (Kingofd) [reply src/units/unit.cpp:815]: You are right, there needs to be a filter for that.

File: src/units/unit.cpp
Comment (Kingofd) [reply src/units/unit.cpp]: It was for an early prototype that would allow for duplicate traits. Needless to say it now is not required, as every trait will only ever be present once on a unit

File: src/units/unit.cpp
Comment (Kingofd) [reply src/units/unit.cpp:862]: I will make sure to include braces, whenever I use if's. I personally find it rather appealing + less cluttering -> easier to read, so changing the code-style might be a nice idea

**Comment by Kingofd** [comment]
I was thinking the same thing. Where should I start with exploring the current test infrastructure?

I guess these test cases should work:

1. Test trait for exclusion ("not with")
2. Test trait for inclusion ("not without")

Anything else that should be included in the new tests?

**Comment by Pentarctagon** [comment]
The current tests are in the `data/test/scenarios/` folder. They are executed by the `run_wml_tests` python script, with the list of tests and expected results in the `wml_test_schedule`. They can be run individually with either `wesnoth -u <test_name>` or `wesnoth -u <test_name> --showgui`.

**Comment by Pentarctagon** [comment]
Tests aren't actually run unless they're added to `wml_test_schedule`.

File: data/core/macros/traits.cfg
Comment (Pentarctagon) [inline data/core/macros/traits.cfg:434]: All of these `#define`s should be part of the unit test cfg file.

File: data/core/units/fake/Test_Unit_Require.cfg
Comment (Pentarctagon) [inline data/core/units/fake/Test_Unit_Require.cfg:24]: Both of these unit types should also be located in the `data/test/` directory somewhere.

Comment (Kingofd) [reply data/core/units/fake/Test_Unit_Require.cfg:24]: I have made a new folder for test units in general. Where do I add them to the environment, the tests can't seem to find them now.

File: data/schema/units/modifications.cfg
Comment (Kingofd) [inline data/schema/units/modifications.cfg:221]: My bad, the units should not have that keyword, it was wrongfully still in the unit file, I'll revert the change. The feature is strictly trait-macro bound, the units should not notice anything.

File: data/core/units/fake/Test_Unit_Require.cfg
Comment (Pentarctagon) [reply data/core/units/fake/Test_Unit_Require.cfg:24]: For regular units, you'd need to add an include for them in https://github.com/wesnoth/wesnoth/blob/master/data/core/units.cfg

For test-only units, I believe you should be able to include them similarly in https://github.com/wesnoth/wesnoth/blob/master/data/test/_main.cfg

Comment (Kingofd) [reply data/core/units/fake/Test_Unit_Require.cfg]: Added test unit capability to test textdomain :+1:

**Comment by CelticMinstrel** [comment]
Is there any remaining reason to hold off on merging this? It seems okay to me.

**Comment by Pentarctagon** [comment]
No objections from me.

File: data/test/_main.cfg
Comment (CelticMinstrel) [inline data/test/_main.cfg:77]: I think the correct fix is instead to move this file into a different folder. @Pentarctagon should be able to help you figure out where it goes.

File: wml_test_schedule
Comment (CelticMinstrel) [inline wml_test_schedule:351]: You need to remove the merge markers here.

**Comment by Pentarctagon** [comment]
`TRAIT_TEST_EXCLUDE`, `TRAIT_TEST_EXCLUDED`, `TRAIT_TEST_REQUIRE`, `TRAIT_TEST_REQUIRED` don't seem to be defined anywhere, so that's why the tests are failing.

**Comment by CelticMinstrel** [comment]
I guess they were accidentally deleted in 584d53bd963c5c33033f979c3c7b900fcbd829bf?

File: data/test/_main.cfg
Comment (Kingofd) [reply data/test/_main.cfg:77]: I was thinking the same thing, where should I put it @Pentarctagon ?

**Comment by Kingofd** [comment]
Hmm, that's weird. I initially put them into the test file, however the test failed with the same error. Now I put it in its own file in `wesnoth/data/test/macros/traits.cfg`, where it should be loaded once the TEST environment is loaded, akin to the main `traits.cfg`.

I guess back to the test file it is (but then it needs to also be available for the `units.cfg`)?
Its also weird because it is a newer error, it did not appear in earlier iterations of this feature.

**Comment by CelticMinstrel** [comment]
I'd also like to request that the new files be renamed to all lower-case.

File: data/test/_main.cfg
Comment (Pentarctagon) [reply data/test/_main.cfg:77]: [trait] is under UnitsWML on the wiki, so it'd be https://github.com/wesnoth/wesnoth/tree/master/data/test/scenarios/wml_tests/UnitsWML

File: data/test/scenarios/traits_exclude_include.cfg
Comment (Pentarctagon) [inline data/test/scenarios/traits_exclude_include.cfg:4]: `exclude [trait] functionality` should be `[trait]exclude_traits=`

File: data/test/scenarios/traits_exclude_include.cfg
Comment (Pentarctagon) [inline data/test/scenarios/traits_exclude_include.cfg:47]: `require [trait] functionality` should be `[trait]require_traits=`

**Comment by Pentarctagon** [comment]
Hmm... WML validation failed again, but logged nothing about why.

**Comment by Kingofd** [comment]
I ran the CI locally, there were some errors with the t_strings in the new test race, they should be fixed now. 

Just FYI, incidentally another error was shown specifically on my local installation: 
`Game Config failed validation! 
20230114 15:01:17 error preprocessor: Found filename containing whitespace: '_initial.cfg' in included directory 'data/schema/./filters'.`

It is not shown in the CI tests so it probably is something on my own copy.

File: data/test/scenarios/wml_tests/UnitsWML/traits_exclude_include.cfg
Comment (Pentarctagon) [inline data/test/scenarios/wml_tests/UnitsWML/traits_exclude_include.cfg:4]: Remove `functionality`.

File: data/test/scenarios/wml_tests/UnitsWML/traits_exclude_include.cfg
Comment (Pentarctagon) [inline data/test/scenarios/wml_tests/UnitsWML/traits_exclude_include.cfg:47]: Remove `functionality`.

File: data/test/units.cfg
Comment (CelticMinstrel) [inline data/test/units.cfg:11]: It's fine to use the original strings but just make them translatable. Since this file declares its textdomain to be `wesnoth-test`, nothing in here will be added to the gettext files, which don't scrape that textdomain. (This also means there's no need to use the `race^` prefix.)

Comment (Kingofd) [reply data/test/units.cfg]: Alright I deleted the prefixes so it should be fine now

**Comment by Kingofd** [comment]
So after this is merged, where would I (or who) update the WML documentation?

**Comment by Pentarctagon** [comment]
It's usually preferred that the person making the change also update the wiki documentation to reflect that change. So you'd update https://wiki.wesnoth.org/UnitsWML#.5Btrait.5D to add the two new attribute along with the note "1.17.13 or later" (similar to the `help_text` attribute).

**Comment by Kingofd** [comment]
So, there are some HTTP error 5xx in the CI builds?

**Comment by stevecotton** [comment]
Sadly we do sometimes get network errors on the CI, I'll rerun them.

**Comment by CelticMinstrel** [comment]
By the way, the email on your commits does not seem to match the email that GitHub has for you, so you may want to either update your git config for the Wesnoth repository or let GitHub know that that's your email.

**Comment by stevecotton** [comment]
Please document the new WML on the wiki.

**Pull request merged**
